open_spiel/games/kriegspiel.h

```diff
@@ -0,0 +1,276 @@
+                 bool rule_50_move);
+  KriegspielState(const KriegspielState&) = default;
+
+  KriegspielState& operator=(const KriegspielState&) = default;
+
+  Player CurrentPlayer() const override {
+    return IsTerminal() ? kTerminalPlayerId : ColorToPlayer(Board().ToPlay());
+  }
+  std::vector<Action> LegalActions() const override;
+  std::string ActionToString(Player player, Action action) const override;
+  std::string ToString() const override;
+
+  bool IsTerminal() const override {
+    return static_cast<bool>(MaybeFinalReturns());
+  }
+
+  std::vector<double> Returns() const override;
+  std::string InformationStateString(Player player) const override;
+  std::string ObservationString(Player player) const override;
+  void ObservationTensor(Player player,
+                         absl::Span<float> values) const override;
+  std::unique_ptr<State> Clone() const override;
+  void UndoAction(Player player, Action action) override;
+
+  // Current board.
+  chess::ChessBoard& Board() { return current_board_; }
+  const chess::ChessBoard& Board() const { return current_board_; }
+  int BoardSize() const { return current_board_.BoardSize(); }
+
+  // Starting board.
+  chess::ChessBoard& StartBoard() { return start_board_; }
+  const chess::ChessBoard& StartBoard() const { return start_board_; }
+
+  std::vector<std::pair<chess::Move, KriegspielUmpireMessage>>& MoveMsgHistory() {
+    return move_msg_history_;
+  }
+  const std::vector<std::pair<chess::Move, KriegspielUmpireMessage>>& MoveMsgHistory() const {
+    return move_msg_history_;
+  }
+
+ protected:
+  void DoApplyAction(Action action) override;
+
+ private:
+
+  friend class KriegspielObserver;
+
+  // Draw can be claimed under the FIDE 3-fold repetition rule (the current
+  // board position has already appeared twice in the history).
+  bool Is3FoldRepetitionDraw() const;
+
+  // Calculates legal actions and caches them. This is separate from
+  // LegalActions() as there are a number of other methods that need the value
+  // of LegalActions. This is a separate method as it's called from
+  // IsTerminal(), which is also called by LegalActions().
+  void MaybeGenerateLegalActions() const;
+
+  absl::optional<std::vector<double>> MaybeFinalReturns() const;
+
+  // We have to store every move made to check for repetitions and to implement
+  // undo. We store the current board position as an optimization.
+  std::vector<std::pair<chess::Move, KriegspielUmpireMessage>> move_msg_history_;
+  // We store this info as an optimisation so that we don't have to compute it
+  // from move_msg_history for observations
+  std::optional<KriegspielUmpireMessage> last_public_msg{};
+  std::optional<KriegspielUmpireMessage> before_last_public_msg{};
+  // Moves that the player tried and were illegal. We don't let player try them
+  // again on the same board because they are clearly still illegal;
+  std::set<chess::Move> illegal_tried_moves_;
```

Comment:
Not sure about this, but I think having a vector and looping through it will be faster than a set, especially when the number of moves is small.